src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,134 @@
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        Node<T> newNode = new Node<>(null,value,null);
+        if (index == 0) {
+            if (head == null) {
+                head = newNode;
+                tail = newNode;
+            } else {
+                newNode.next = head;
+                head.prev = newNode;
+                head = newNode;
+            }
+        } else if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = findByIndex(index);
+            current.prev.next = newNode;
+            newNode.prev = current.prev;
+            newNode.next = current;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findByIndex(index);
+        T changedValue = currentNode.item;
+        currentNode.item = value;
+        return changedValue;
     }
 
     @Override
     public T remove(int index) {
+        checkIndex(index);
+        Node<T> currentNode = findByIndex(index);
+        currentNode.prev = currentNode.next;
+        currentNode.next.prev = currentNode.prev;
+        size--;
         return null;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            Node<T> nodeToRemove = findByIndex(i);
+            if ((object == null && nodeToRemove.item == null)
+                    || (object != null && object.equals(nodeToRemove.item))) {
+                nodeToRemove.prev = nodeToRemove.next;
+                nodeToRemove.next.prev = nodeToRemove.prev;
+                size--;
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size <= 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't found");
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't found");
+        }
+    }
+
+    private Node<T> findByIndex(int index) {
+        Node<T> currentByIndex;
+        currentByIndex = head;
+        for (int i = 1; i < size; i++) {
```

Comment:
The `findByIndex` method is not iterating nodes efficiently. You should iterate only the first or second half of the list, depending on the index value, to improve performance.